### PR TITLE
Version 2

### PR DESCRIPTION
**Pull request by cniska**
Bootstrap project with Create React App instead of using custom configuration.

### Todo before this can be merged

- [x] Add and configure Stylelint @crisu83
- [x] Ensure that HMR is working correctly @AriaMinaei 
- [x] Ensure that Flow is properly configured @AriaMinaei 
- [x] Rewrite tests with Jest (try codemods) @crisu83
- [x] Upgrade to React Router v4 @crisu83
- [x] Add Reselect (and rewrite selectors) @crisu83
- [x] Go through dependencies
- [x] Update documentation

### Things you might be wondering about

- Why no `es2015` or `stage-1` Babel presets? Adding these presets has severe performance implications, so we wanted to leave it up to the project to decide which presets it actually needs. See also https://github.com/facebookincubator/create-react-app/issues/435
- Why [Jest](https://facebook.github.io/jest/)? Facebook is using Jest for testing all of their JavaScript and they have made huge improvements to it during the past year. The DE (Developer Experience) is also superior to any other testing framework).
- Why [Reselect](https://github.com/reactjs/reselect)? We have noticed that we need a unified way to select data from the state to ease the work of the developer. Reselect has detailed documentation and it may also increase the performance of the application with memorized selectors.

### CODE REVIEW

**Comment by hugovk** [comment]
CI is failing:
```
$ flow 
sh: 1: flow: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "flow" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by cniska** [comment]
I fixed the build, but I still need to go over all the changes and verify that everything is OK.

**Comment by cniska** [comment]
@AriaMinaei @ericnishio @Kotpes @villeristi could you also please go through the changes and approve them? Thank you in advance.

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,26 @@
+<<<<<<< HEAD
```
Comment (Jalle19) [inline .gitignore]: Merge error

File: .lintstagedrc [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   "*.js": [
+    "git add .",
```
Comment (Jalle19) [inline .lintstagedrc]: But why?

File: README.md [hunk 3]
```diff
@@ -10,7 +10,11 @@
 
 ## Why do I want this?
 
+<<<<<<< HEAD
```
Comment (Jalle19) [inline README.md]: Lots of merge errors here

File: package.json [hunk 4]
```diff
@@ -1,83 +1,109 @@
...
+    "normalize.css": "^6.0.0",
+    "react": "^15.5.4",
+    "react-dom": "^15.5.4",
+    "react-redux": "^5.0.4",
+    "react-router": "3",
+    "redux": "^3.6.0",
+    "redux-actions": "^2.0.2",
+    "redux-saga": "^0.14.7"
+  },
+  "devDependencies": {
```
Comment (Jalle19) [inline package.json]: IMO everything that is not related to `webpack-dev-server`, linting etc. should be in `dependencies`.

File: .gitignore (hunk 1 above)
Comment (cniska) [reply .gitignore]: Will fix, this is basically a completely new code base that was rebased on `master`.

Comment (crisu83) [inline .gitignore]: Will fix, this is basically a completely new code base that was rebased on `master`.

File: package.json (hunk 4 above)
Comment (cniska) [reply package.json]: But why? We should be installing dev dependencies during CI and only leave them out when building production.

File: .lintstagedrc (hunk 2 above)
Comment (cniska) [reply .lintstagedrc]: Because it's one less manual step that speeds up development and makes people work in smaller chunks. https://github.com/okonet/lint-staged#examples

Comment (Jalle19) [reply .lintstagedrc]: Think about this scenario:

1. You work on something
2. You've staged parts of some files and you're ready to commit
3. You run the linter, and `git add .` fucks up your stage. Now you have to unstage everything and carefully stage the lines/chunks/files you had before.

File: package.json (hunk 4 above)
Comment (Jalle19) [reply package.json]: We've discussed this once already. The fact is that there are no "production dependencies" since the production build is just a bunch of `.js` files and some assets.

Currently in order to build the production bundle you have to install every single dependency, including `webpack-dev-server`.

You have to ask yourself "what can I do if I run `yarn install --production`"? With the current `package.json` you will not be able to do anything, not even, i.e. none of the commands here will work (AFAIK):

```
+  "scripts": {
+    "ci": "CI=true run-p flow lint-staged test",
+    "flow": "flow",
+    "lint": "eslint src --fix",
+    "lint-staged": "lint-staged",
+    "precommit": "lint-staged",
+    "stylelint": "stylelint 'src/**/*.css'",
+    "start": "node scripts/start.js",
+    "build": "node scripts/build.js",
+    "test": "node scripts/test.js --env=jsdom"
+  },
```

File: .lintstagedrc (hunk 2 above)
Comment (hugovk) [reply .lintstagedrc]: I'm also a bit dubious. Did this cause the .gitignore merge errors to be committed? If anything, perhaps it should only add the modified .js files.

 https://github.com/okonet/lint-staged#examples says:

> Please note, that it doesn’t work well with committing hunks (`git add -p`).

Comment (Jalle19) [reply .lintstagedrc]: Indeed, IMO staging/committing is a job for the developer, not the machine.

**Review by Kotpes — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -61,10 +71,10 @@ yarn start
 
 ### Distribution build
 
-You can compile the distribution build with the `compile` script.
+You can compile the distribution build with the `build` script.
 
 ```bash
-yarn run compile
+yarn run build
```
Comment (ericnishio) [inline README.md]: Or simply `yarn build`.

File: config/env.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+function getClientEnvironment(publicUrl) {
+  var raw = Object
+    .keys(process.env)
+    .filter(key => REACT_APP.test(key))
+    .reduce((env, key) => {
+      env[key] = process.env[key];
+      return env;
+    }, {
+      // Useful for determining whether we’re running in production mode.
```
Comment (ericnishio) [inline config/env.js:16]: Overall, I think the inline documentation is too verbose.

File: config/paths.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+var path = require('path');
```
Comment (ericnishio) [inline config/paths.js:3]: Since we're no longer supporting older versions of Node, we should be able to replace all instances of `var` with `const` (and `let` if needed).

File: config/polyfills.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+'use strict';
```
Comment (ericnishio) [inline config/polyfills.js:1]: Is this needed?

File: config/webpack.config.dev.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+var webpack = require('webpack');
+var HtmlWebpackPlugin = require('html-webpack-plugin');
+var CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');
+var InterpolateHtmlPlugin = require('react-dev-utils/InterpolateHtmlPlugin');
+var WatchMissingNodeModulesPlugin = require('react-dev-utils/WatchMissingNodeModulesPlugin');
+var getClientEnvironment = require('./env');
+var paths = require('./paths');
+
+
+
```
Comment (ericnishio) [inline config/webpack.config.dev.js:16]: Too many empty lines.

File: config/env.js (hunk 6 above)
Comment (cniska) [reply config/env.js:16]: It's all from Create React App with as few modifications as possible.

**Comment by cniska** [comment]
@Jalle19 let's look at the dependencies today when we meet face to face.

File: src/index.js [hunk 10]
```diff
@@ -1,4 +1,33 @@
-import startApp from './root/startApp'
+// @flow
+
+import React, {createElement} from 'react'
+import {render} from 'react-dom'
+import {AppContainer} from 'react-hot-loader'
+import {browserHistory} from 'react-router'
```
Comment (hugovk) [inline src/index.js]: > 'browserHistory' is defined but never used.

**Comment by Jalle19** [comment]
+1 from me as long as `rm -rf node_modules && yarn install --production && yarn run compile` works

File: .gitignore [hunk 11]
```diff
@@ -1,5 +1,18 @@
-/coverage/
-/node_modules/
-/dist/
+# See https://help.github.com/ignore-files/ for more about ignoring files.
+
```
Comment (AriaMinaei) [inline .gitignore:2]: Why are we committing the build folder to the repo? @crisu83

Comment (cniska) [reply .gitignore:2]: We are not?
